Enable client side digest pinning for stack deploy

nishanttotla opened the pull request:
This PR is a follow-up to https://github.com/docker/cli/pull/30.

It also adds a `--no-resolve-image` flag to `stack deploy` to disable digest pinning if required.

I'm not sure if we want to do this for 17.06, but it seems like all the required functionality exists, and it would be useful to have consistency.

cc @dnephin @tiborvass  @thaJeztah

aaronlehmann commented:
I think we need this for 17.06 because if we didn't have it, we'd lose digest pinning from `stack deploy`, and that would be a regression.

nishanttotla commented:
@dnephin @tiborvass what can we do to make sure this makes it into 17.06?

Also opened https://github.com/moby/moby/pull/33386 to ensure tests don't contact registry because of this.

aaronlehmann commented on cli/command/stack/deploy_composefile.go:
```diff
@@ -320,6 +322,13 @@ func deployServices(
 				updateOpts.EncodedRegistryAuth = encodedAuth
 			}
+
+			if image != service.Spec.TaskTemplate.ContainerSpec.Image {
```
I'm not sure this check makes sense. It means that if the digest wasn't pinned before, we won't pin the digest (or add platform information) going forward. Something like a temporary registry error would cause pinning to be turned off for future runs as well.

nishanttotla replied:
@aaronlehmann you're right, that would be a problem.

For `service update`, it's pretty simple to judge because we have a `--image` flag, but I couldn't find a way of deciding whether an image was updated in an updated stack file. Is there a way of discovering this at the CLI level? cc @dnephin

aaronlehmann replied:
I think it would be best to just remove the check, so the digest and platform information are always updated.

nishanttotla replied:
I'm a little uncomfortable with that idea, because if the stack file contains multiple images, then it's possibly too much overhead each time an update is done.

Although it's easy to avoid that given the `--no-resolve-image` flag, but the some might find that frustrating UI. Anyhow, if there's no other way, then I agree that removing the check is the right thing to do.

aaronlehmann replied:
I don't see a better way, unless we want to change the UX to make digest pinning a manual operation, rather than getting the latest digest on every deploy.

aaronlehmann replied:
That sounds reasonable to me.

We would need to make sure there is a way to enable digest pinning if the service was originally deployed with `--no-resolve-image`, or vice versa.

dnephin replied:
Do we expect `--no-resolve-image` to persist? I expect that it would only apply to the current deploy.

For any service with a missing label we can treat that as an image name mismatch. The first deploy will cause an update to include the service image label. 

Otherwise I wouldn't expect `--no-resolve-image` to change the label value (it will always contain the string from the config, not a digest), so the flag acts as an override.

aaronlehmann replied:
I wouldn't expect it to persist, but just saying that storing the original name in a label and comparing with that when deciding whether to update the digest could cause us not to add a digest if there wasn't one originally. We just need to make sure to take that account in the logic.

nishanttotla replied:
It isn't very clear to me how having a label solves the problem. If the label contains the image that was originally deployed (without a digest because say pinning failed), then later when the stack file is updated (without `--no-resolve-image` flag, but the same image), how do we decide to contact the registry?

To me, the `--image` flag in `service create` and `service update` also conveys "user intent", something that I'm not able to grasp in stack files. Please correct me if I've misunderstood.

dnephin replied:
The label can give us the same signal as the `--image` flag on service commands.

Here's how it works:
* in `convert.Services()` we add a label to each service with a key of `com.docker.stack.image` (or something like this) and a value of the literal string from the compose file (`composetypes.Service.Image`)
* when we enter this `deployServices()` code we compare the existing label value against the new value from the compose file

```go
image := serviceSpec.TaskTemplate.ContainerSpec.Image
// has the image value changed in the compose file?
if service.Spec.Labels["com.docker.stack.image"] != image {
    if !noResolveImage {
        updateOpts.QueryRegistry = true
    }
}
```

This is different from the current code because we don't know if `service.Spec.TaskTemplate.ContainerSpec.Image` was modified or not, but we do know the label is never modified by other code.

Does that make sense?

dnephin commented on cli/command/stack/deploy_composefile.go:
```diff
@@ -341,6 +350,12 @@ func deployServices(
+			// query registry if flag disabling it was not set
+			if !noResolveImage && versions.GreaterThanOrEqualTo(apiClient.ClientVersion(), "1.30") {
+				createOpts.QueryRegistry = true
+			}
```
Instead of checking the version twice I think we should handle the version check in `runDeploy()` by setting the option accordingly (set the option to` true` for older versions). That way `noResolveImage` will always be correct inside `deployServices()` and we won't need to check version twice.

nishanttotla replied:
@dnephin I addressed this comment, and the one about refactoring.

dnephin commented on cli/command/stack/opts.go:
```diff
@@ -24,6 +24,12 @@ func addRegistryAuthFlag(opt *bool, flags *pflag.FlagSet) {
+	var flagNoResolveImage = "no-resolve-image"
+	flags.BoolVar(opt, flagNoResolveImage, false, "Do not query the registry to resolve image digest and supported platforms")
+	flags.SetAnnotation(flagNoResolveImage, "version", []string{"1.30"})
+}
```
A minor nit: typically we only use these functions for flags that are shared between commands. I think it's fine to keep this as a separate function, but maybe move it to the single file where it's used?

dnephin commented on cli/command/stack/deploy.go:
```diff
@@ -44,12 +46,19 @@ func newDeployCommand(dockerCli command.Cli) *cobra.Command {
 func runDeploy(dockerCli command.Cli, opts deployOptions) error {
 	ctx := context.Background()
 
+	// image resolution should not happen for clients older than v1.30
```
Even though this says "ClientVersion()" it's important to note it actually reflects the supported version of the server. The comment might be a bit misleading in that sense, but I guess it's fine.

aaronlehmann replied:
Yes, I agree with changing this comment.

aaronlehmann commented:
> The label can give us the same signal as the --image flag on service commands.
> 
> Here's how it works:
> 
> in convert.Services() we add a label to each service with a key of com.docker.stack.image (or something like this) and a value of the literal string from the compose file (composetypes.Service.Image)
> when we enter this deployServices() code we compare the existing label value against the new value from the compose file

What about the case where digest pinning fails, and I want to retry it? I would expect rerunning docker stack deploy without --no-resolve-image to do this, but if I understand the proposal correctly, it would detect that the image has not changed in the compose file, and avoid digest pinning.

dnephin commented:
> What about the case where digest pinning fails,

If a service create/update is performed with with digest pinning enabled, but it failed to pin, I would expect the service create/update call to fail, not for the command to return success anyway. Is that not the case?

aaronlehmann commented:
It completes successfully but warns the user that warns the user that the image is referenced by name, not by digest.

Otherwise, you would need to use a special flag to create a service using a local or pre-pulled image, which would break backward compat.

dnephin commented:
I see. In that case I guess we need a way to say "force pinning".  We should still add the label for the default case.  For the force case I think we should change the flag to `--resolve-image=<value>` where values can be (something like): `resolve` (the default), `never` (equivalent to current flag), `force` set `QueryRegistry` even if we can't detect any changes.

aaronlehmann commented:
SGTM

aaronlehmann commented:
> As far as labels go though, consider this example.

This is why I think it's much better to (at least by default) always query the registry. Doing so matches the old behavior when this was done daemon-side, assuming `stack deploy` set the `Image` field in `ContainerSpec` to the value from the compose file instead of leaving the old, pinned value.

dnephin commented:
I'm really not sure what someone would expect when they use `latest`. It's not pinned, so some people might expect it to update, and others might not. 

Maybe the default should be `force` (rename to `always` ?), and `resolve` (rename to `changed`?) could be used when you only way to query if the image name is changed in the file?

nishanttotla commented:
@dnephin I'm possibly restating what's been said above, but just for clarity: changing the default to `force` (or `always`, which sounds more appropriate) sounds equivalent to what @aaronlehmann is suggesting. Then the `resolve` (or `changed`) and 'never' flags will give users more control over that behavior.

With this proposal, digest pinning always happens, unless the user chooses `resolve` in which case it'll happen only when the image provided in the compose file does not match that which is actually running. In this scenario, if digest pinning failed to begin with, it will never be tried again unless `force`d. If digest pinning succeeded with `resolve` in the first place, then it'll still happen if the user provided only a tag.

I'd be satisfied with this sort of behavior, if we all agree on this. (I'll change the option names too)

dnephin commented:
Yes, I think that sounds right.

> In this scenario, if digest pinning failed to begin with, it will never be tried again unless forced

I don't understand this part. Why would this be the case? I think using a label fixes this problem?

nishanttotla commented:
@dnephin my comment above explains in more detail what I mean: https://github.com/docker/cli/pull/121#issuecomment-305928062

nishanttotla commented:
Changed default to `always`.

dnephin commented on cli/compose/convert/service.go:
```diff
@@ -46,6 +46,12 @@ func Services(
+		if serviceSpec.Labels == nil {
+			serviceSpec.Labels = make(map[string]string)
+		}
+		serviceSpec.Labels["com.docker.stack.image"] = service.Image
```
We already add labels in `convertService()`, we should probably keep this together with the other labels we're adding.

A constant for this label would be good as well.

dnephin commented on cli/command/stack/deploy.go:
```diff
@@ -44,12 +46,19 @@ func newDeployCommand(dockerCli command.Cli) *cobra.Command {
+	// image resolution should not happen for clients older than v1.30
+	if versions.LessThan(dockerCli.Client().ClientVersion(), "1.30") {
+		opts.resolveImage = "never"
+	}
```
Right now nothing is validating that the string value of `opts.resolveImage` is one of the 3 valid options. Maybe this check could be moved to a new function along with the validation?

nishanttotla replied:
Should the command error out if the option is invalid?

dnephin replied:
Yes, I think it should error with "invalid option for resolve-image ..." or something like that

nishanttotla commented:
@dnephin added validation for the CLI option, and moved label addition to `convertService()`.

aaronlehmann commented on cli/command/stack/deploy.go:
```diff
@@ -62,6 +71,18 @@ func runDeploy(dockerCli command.Cli, opts deployOptions) error {
 
+// checkImageResolutionAllowed validates the opts.resolveImage command line option
+// and also turns image resolution off if the version is older than 1.30
+func checkImageResolutionAllowed(dockerCli command.Cli, opts *deployOptions) error {
```
`validateResolveImageFlag`?

aaronlehmann commented on cli/command/stack/deploy.go:
```diff
@@ -62,6 +71,18 @@ func runDeploy(dockerCli command.Cli, opts deployOptions) error {
+// checkImageResolutionAllowed validates the opts.resolveImage command line option
+// and also turns image resolution off if the version is older than 1.30
+func checkImageResolutionAllowed(dockerCli command.Cli, opts *deployOptions) error {
+	if opts.resolveImage != "always" && opts.resolveImage != "changed" && opts.resolveImage != "never" {
```
This could be a `switch`, but it's fine this way too.

aaronlehmann replied:
I'd suggest declaring constants for the possible values, so it is harder for this to be out of sync with the code that interprets the flag values, or for typos to creep into the string literals in some places. Again, not a big deal either way.

dnephin commented on cli/compose/convert/service.go:
```diff
@@ -157,6 +161,12 @@ func convertService(
+	// add an image label to serviceSpec
+	if serviceSpec.Labels == nil {
+		serviceSpec.Labels = make(map[string]string)
+	}
```
This condition is already handled by `AddStackLabel()`.

nishanttotla commented:
I've addressed comments and fixed the merge conflict.

For the test, it seems like `deployServices` is the function that would be ideal to test, but `fakeClient` seems to be an API client, and `deployServices` needs to use a `command.Cli` object. Is there one that exists for this purpose?

Alternatively, @dnephin would you suggest a different unit test? Could we make this a follow-up?

nishanttotla commented:
@dnephin it is time sensitive since it must be cherry picked for rc3, so I can add the test later.

nishanttotla commented:
I've confirmed these changes work as desired in local (manual) testing, and opened a cherry-pick PR here: https://github.com/docker/docker-ce/pull/44
